nav2_collision_monitor/src/collision_monitor_node.cpp

```diff
@@ -0,0 +1,535 @@
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "nav2_collision_monitor/collision_monitor_node.hpp"
+
+#include <exception>
+#include <cmath>
+#include <limits>
+#include <utility>
+
+#include "tf2_ros/create_timer_ros.h"
+
+#include "nav2_util/node_utils.hpp"
+
+namespace nav2_collision_monitor
+{
+
+CollisionMonitorNode::CollisionMonitorNode(const rclcpp::NodeOptions & options)
+: nav2_util::LifecycleNode("collision_monitor_node", "", false, options),
+  velocity_({0.0, 0.0, 0.0}), velocity_valid_(false), cmd_vel_in_({0.0, 0.0, 0.0}),
+  ra_prev_({DO_NOTHING, {0.0, 0.0, 0.0}}), release_operation_(false)
+{
+  RCLCPP_INFO(get_logger(), "Creating CollisionMonitorNode");
+}
+
+CollisionMonitorNode::~CollisionMonitorNode()
+{
+  RCLCPP_INFO(get_logger(), "Destroying CollisionMonitorNode");
+  polygons_.clear();
+  sources_.clear();
+}
+
+nav2_util::CallbackReturn
+CollisionMonitorNode::on_configure(const rclcpp_lifecycle::State & /*state*/)
+{
+  RCLCPP_INFO(get_logger(), "Configuring");
+
+  // Transform buffer and listener initialization
+  tf_buffer_ = std::make_shared<tf2_ros::Buffer>(this->get_clock());
+  auto timer_interface = std::make_shared<tf2_ros::CreateTimerROS>(
+    this->get_node_base_interface(),
+    this->get_node_timers_interface());
+  tf_buffer_->setCreateTimerInterface(timer_interface);
+  tf_listener_ = std::make_shared<tf2_ros::TransformListener>(*tf_buffer_); 
+
+  // Obtaining ROS parameters
+  if (!getParameters()) {
+    return nav2_util::CallbackReturn::FAILURE;
+  }
+
+  // Setting subscribers and publishers
+  rclcpp::QoS odom_qos = rclcpp::SystemDefaultsQoS();  // set to default
+  odom_sub_ = this->create_subscription<nav_msgs::msg::Odometry>(
+    odom_topic_, odom_qos,
+    std::bind(&CollisionMonitorNode::odomCallback, this, std::placeholders::_1));
+
+  cmd_vel_in_sub_ = this->create_subscription<geometry_msgs::msg::Twist>(
+    cmd_vel_in_topic_, 1,
+    std::bind(&CollisionMonitorNode::cmdVelInCallback, this, std::placeholders::_1));
+  cmd_vel_out_pub_ = this->create_publisher<geometry_msgs::msg::Twist>(
+    cmd_vel_out_topic_, 1);
+
+  if (polygon_topic_.length()) {
+    rclcpp::QoS polygon_qos = rclcpp::SystemDefaultsQoS();  // set to default
+    polygon_pub_ = this->create_publisher<geometry_msgs::msg::PolygonStamped>(
+      polygon_topic_, polygon_qos);
+  }
+
+  return nav2_util::CallbackReturn::SUCCESS;
+}
+
+nav2_util::CallbackReturn
+CollisionMonitorNode::on_activate(const rclcpp_lifecycle::State & /*state*/)
+{
+  RCLCPP_INFO(get_logger(), "Activating");
+
+  // Activating lifecycle publishers
+  polygon_pub_->on_activate();
+  cmd_vel_out_pub_->on_activate();
+
+  // Activating main worker
+  setWorkerActive(true);
+
+  // Creating bond connection
+  createBond();
+
+  // Polygons publishing thread
+  polygon_pub_timer_ = this->create_wall_timer(
```

Comment:
Remove and publish in the main thread if `visualize` is true (default false)

Reply:
I've prepared the update with publishing polygons from main thread. However, w/o a timer main processing routine works only when input `cmd_vel` messages are arriving. This means, if we will start robot simulation, at the beginning it will stand, no `cmd_vel` messages will appear and polygons are not being visualized. They are being appear on RViz screen only after robot start moving.

I've tried to solve the problem by publishing polygons once at `on_activate()` stage with transient local durability, but RViz still can't see initially published polygon topic. Tomorrow, will try to work on this more, but now I have only one solution - is to keep separate timer for polygon publisher.

Reply:
I think the activate publication is smart, but you don't need transient local durability, that is probably why you don't see it in rviz - rviz also needs its QoS to be tuned. If rviz is already up, then it'll get the message

Reply:
The polygon did not appear in RViz on CM start only if I add it "on-the-fly", when RViz is already running. I've used transient_local durability for RViz, and both QoS modes publishing for CM node. However, if I will save the RViz config with added polygon; next time when RViz will already load the config, polygon will appear right after CM start, as needed. This most probably it is an RViz issue, so I am leaving polygons publishing in a main loop + in `on_activate()` stage for now.